{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import necessary libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Set the path for the data\n",
    "input_path = '/mnt/Local_Disk_1/2_Hospital_Microbiome/Data/Output_data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy ID</th>\n",
       "      <th>Pathogen name</th>\n",
       "      <th>Host</th>\n",
       "      <th>Disease</th>\n",
       "      <th>NO. of MLSTs</th>\n",
       "      <th>NO. of strains</th>\n",
       "      <th>NO. of genomes</th>\n",
       "      <th>Biosafety level</th>\n",
       "      <th>NO. of resistant drugs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>Birds,Environment,Food,Human,Mammal,Primate,Ro...</td>\n",
       "      <td>Blood,Bone/Joint,Cardiovascular system,Central...</td>\n",
       "      <td>4678</td>\n",
       "      <td>503650</td>\n",
       "      <td>548703</td>\n",
       "      <td>BSL-2</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Birds,Environment,Food,Human,Mammal,Mollusc,Pr...</td>\n",
       "      <td>Blood,Bone/Joint,Cardiovascular system,Central...</td>\n",
       "      <td>6340</td>\n",
       "      <td>219751</td>\n",
       "      <td>282504</td>\n",
       "      <td>BSL-2</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1280</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>Birds,Environment,Food,Human,Mammal,Primate,Ro...</td>\n",
       "      <td>Blood,Bone/Joint,Cardiovascular system,Central...</td>\n",
       "      <td>1628</td>\n",
       "      <td>55519</td>\n",
       "      <td>92673</td>\n",
       "      <td>BSL-2</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>197</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Birds,Human,Mammal,Primate,Rodent</td>\n",
       "      <td>Blood,Digestive system,Urinary system</td>\n",
       "      <td>2009</td>\n",
       "      <td>68479</td>\n",
       "      <td>81398</td>\n",
       "      <td>BSL-2</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Birds,Environment,Food,Human,Mammal,Primate,Ro...</td>\n",
       "      <td>Blood,Bone/Joint,Cardiovascular system,Central...</td>\n",
       "      <td>3443</td>\n",
       "      <td>51133</td>\n",
       "      <td>75987</td>\n",
       "      <td>BSL-2</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Taxonomy ID          Pathogen name  \\\n",
       "0        28901    Salmonella enterica   \n",
       "1          562       Escherichia coli   \n",
       "2         1280  Staphylococcus aureus   \n",
       "3          197   Campylobacter jejuni   \n",
       "4          573  Klebsiella pneumoniae   \n",
       "\n",
       "                                                Host  \\\n",
       "0  Birds,Environment,Food,Human,Mammal,Primate,Ro...   \n",
       "1  Birds,Environment,Food,Human,Mammal,Mollusc,Pr...   \n",
       "2  Birds,Environment,Food,Human,Mammal,Primate,Ro...   \n",
       "3                  Birds,Human,Mammal,Primate,Rodent   \n",
       "4  Birds,Environment,Food,Human,Mammal,Primate,Ro...   \n",
       "\n",
       "                                             Disease  NO. of MLSTs  \\\n",
       "0  Blood,Bone/Joint,Cardiovascular system,Central...          4678   \n",
       "1  Blood,Bone/Joint,Cardiovascular system,Central...          6340   \n",
       "2  Blood,Bone/Joint,Cardiovascular system,Central...          1628   \n",
       "3              Blood,Digestive system,Urinary system          2009   \n",
       "4  Blood,Bone/Joint,Cardiovascular system,Central...          3443   \n",
       "\n",
       "   NO. of strains  NO. of genomes Biosafety level  NO. of resistant drugs  \n",
       "0          503650          548703           BSL-2                      30  \n",
       "1          219751          282504           BSL-2                      28  \n",
       "2           55519           92673           BSL-2                      27  \n",
       "3           68479           81398           BSL-2                      25  \n",
       "4           51133           75987           BSL-2                      37  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finding pathogen\n",
    "# Loading pathogen data\n",
    "pathogen_data = pd.read_csv(input_path + 'gcPathogen_20241019.txt', sep = '|')\n",
    "\n",
    "pathogen_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Taxonomy ID', 'Pathogen name', 'Host', 'Disease', 'NO. of MLSTs',\n",
       "       'NO. of strains', 'NO. of genomes', 'Biosafety level',\n",
       "       'NO. of resistant drugs'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathogen_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hospital</th>\n",
       "      <th>MetaSUB</th>\n",
       "      <th>Office</th>\n",
       "      <th>Present</th>\n",
       "      <th>Biosafety level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abiotrophia defectiva</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>BSL-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Achromobacter xylosoxidans</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acinetobacter baumannii</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>BSL-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acinetobacter johnsonii</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>BSL-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acinetobacter junii</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>BSL-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stutzerimonas stutzeri</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Veillonella atypica</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>BSL-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Veillonella parvula</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Xanthomonas citri</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>[Ruminococcus] torques</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>No</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Hospital  MetaSUB  Office Present Biosafety level\n",
       "Abiotrophia defectiva              1        0       1     Yes           BSL-2\n",
       "Achromobacter xylosoxidans         1        0       0      No              NA\n",
       "Acinetobacter baumannii            1        0       0     Yes           BSL-2\n",
       "Acinetobacter johnsonii            1        1       0     Yes           BSL-2\n",
       "Acinetobacter junii                1        0       1     Yes           BSL-2\n",
       "...                              ...      ...     ...     ...             ...\n",
       "Stutzerimonas stutzeri             1        1       1      No              NA\n",
       "Veillonella atypica                1        0       0     Yes           BSL-1\n",
       "Veillonella parvula                1        1       0     Yes              NA\n",
       "Xanthomonas citri                  1        0       0      No              NA\n",
       "[Ruminococcus] torques             1        0       0      No              NA\n",
       "\n",
       "[114 rows x 5 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the species data\n",
    "taxa_data = pd.read_csv(input_path + 'taxa_binary.csv', sep = ',', index_col=0)\n",
    "\n",
    "taxa_data.head()\n",
    "taxa_data['Present'] = taxa_data.index.to_series().apply(lambda x: 'Yes' if any(pathogen_data['Pathogen name'] == x) else 'No')\n",
    "taxa_data['Biosafety level'] = taxa_data.index.to_series().apply(lambda x: pathogen_data[pathogen_data['Pathogen name'] == x]['Biosafety level'].values[0] if any(pathogen_data['Pathogen name'] == x) else None)\n",
    "taxa_data['Biosafety level'].replace(np.nan, 'NA', inplace=True)\n",
    "taxa_data['Biosafety level'].fillna('NA', inplace=True)\n",
    "taxa_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the data\n",
    "taxa_data.to_csv(input_path + 'taxa_binary_pathogen.tsv', sep = '\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
