{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import necessary packages\n",
    "import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Set the directory path\n",
    "figure_path = '/mnt/Local_Disk_1/Hospital_Microbiome/Data/Figures/'\n",
    "input_path = '/mnt/Local_Disk_1/Hospital_Microbiome/Data/Input_data/'\n",
    "model_path = '/mnt/Local_Disk_1/Hospital_Microbiome/Data/Modeling/Models/'\n",
    "output_path = '/mnt/Local_Disk_1/Hospital_Microbiome/Data/Output_data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>gram_staining</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organism_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abiotrophia defectiva</th>\n",
       "      <td>GCF_037041345.1</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Achromobacter xylosoxidans</th>\n",
       "      <td>GCF_022870085.1</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acinetobacter baumannii</th>\n",
       "      <td>GCF_000369385.1</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acinetobacter johnsonii</th>\n",
       "      <td>GCF_016027055.1</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acinetobacter junii</th>\n",
       "      <td>GCF_000430225.1</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stutzerimonas stutzeri</th>\n",
       "      <td>GCF_000219605.1</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Veillonella atypica</th>\n",
       "      <td>GCF_002959915.1</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Veillonella parvula</th>\n",
       "      <td>GCF_000024945.1</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Xanthomonas citri</th>\n",
       "      <td>GCF_002018575.1</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>[Ruminococcus] torques</th>\n",
       "      <td>GCF_000153925.1</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           assembly_accession gram_staining\n",
       "organism_name                                              \n",
       "Abiotrophia defectiva         GCF_037041345.1      positive\n",
       "Achromobacter xylosoxidans    GCF_022870085.1      negative\n",
       "Acinetobacter baumannii       GCF_000369385.1      negative\n",
       "Acinetobacter johnsonii       GCF_016027055.1      negative\n",
       "Acinetobacter junii           GCF_000430225.1      negative\n",
       "...                                       ...           ...\n",
       "Stutzerimonas stutzeri        GCF_000219605.1      negative\n",
       "Veillonella atypica           GCF_002959915.1      negative\n",
       "Veillonella parvula           GCF_000024945.1      negative\n",
       "Xanthomonas citri             GCF_002018575.1      negative\n",
       "[Ruminococcus] torques        GCF_000153925.1      positive\n",
       "\n",
       "[114 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the gram staining information file\n",
    "staining_info = pd.read_csv(input_path + 'gram_staining.csv', sep=',', index_col=\"organism_name\")\n",
    "staining_info\n",
    "\n",
    "# Load the genome information file\n",
    "genome_info = pd.read_csv(output_path + 'genome_details.csv', sep=',', index_col=\"organism_name\")\n",
    "genome_info = genome_info[[\"assembly_accession\"]]\n",
    "\n",
    "# Matching index\n",
    "genome_info.index = genome_info.index.to_series().apply(lambda x: ' '.join(x.split()[:2]))\n",
    "\n",
    "# Merge the two dataframes\n",
    "genome_info = genome_info.merge(staining_info, left_index=True, right_index=True)\n",
    "\n",
    "genome_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the accession id list\n",
    "accessions = genome_info['assembly_accession'].to_list()\n",
    "\n",
    "# Specify the file path\n",
    "accession_path = output_path + 'accession.list'\n",
    "\n",
    "# Write the list to a text file\n",
    "with open(accession_path, 'w') as f:\n",
    "    for accession in accessions:\n",
    "        f.write(\"%s\\n\" % accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Performed in the terminal\n",
    "# conda activate bit\n",
    "# bit-dl-ncbi-assemblies -w accession.list -j 80 -f fasta\n",
    "# gunzip *.gz\n",
    "\n",
    "# Run prokka\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated bash commands written to /mnt/Local_Disk_1/Hospital_Microbiome/Data/Modeling/Models/generate_models.sh\n"
     ]
    }
   ],
   "source": [
    "carve_file_path = model_path + 'generate_models.sh'\n",
    "annotated_file_path = '/mnt/Local_Disk_1/Hospital_Microbiome/Data/Modeling/Annotated_genomes/'\n",
    "with open(carve_file_path, \"w\") as f:\n",
    "    f.write(\"#!/bin/bash\\n\")\n",
    "    for idx, row in genome_info.iterrows():\n",
    "        f.write(f\"carve {annotated_file_path}{row['assembly_accession']}.faa -o {idx.replace(' ', '_')}.xml --solver cplex -u {'grampos' if row['gram_staining'] == 'positive' else 'gramneg'}\\n\")\n",
    "        f.write(f\"echo {idx} model is done\\n\")\n",
    "# Print the completion message\n",
    "print(f\"Generated bash commands written to {carve_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Abiotrophia defectiva model is done\n",
      "Achromobacter xylosoxidans model is done\n",
      "Acinetobacter baumannii model is done\n",
      "Acinetobacter johnsonii model is done\n",
      "Acinetobacter junii model is done\n",
      "Acinetobacter lwoffii model is done\n",
      "Acinetobacter parvus model is done\n",
      "Acinetobacter schindleri model is done\n",
      "Acinetobacter ursingii model is done\n",
      "Actinomyces oris model is done\n",
      "Actinomyces viscosus model is done\n",
      "Aerococcus viridans model is done\n",
      "Bacteroides thetaiotaomicron model is done\n",
      "Bacteroides uniformis model is done\n",
      "Bifidobacterium adolescentis model is done\n",
      "Bifidobacterium breve model is done\n",
      "Bifidobacterium longum model is done\n",
      "Brachybacterium muris model is done\n",
      "Brachybacterium paraconglomeratum model is done\n",
      "Brevibacterium casei model is done\n",
      "Brevibacterium ravenspurgense model is done\n",
      "Brevundimonas diminuta model is done\n",
      "Caulobacter vibrioides model is done\n",
      "Chroococcidiopsis thermalis model is done\n",
      "Collinsella aerofaciens model is done\n",
      "Corynebacterium accolens model is done\n",
      "Corynebacterium durum model is done\n",
      "Corynebacterium glucuronolyticum model is done\n",
      "Corynebacterium jeikeium model is done\n",
      "Corynebacterium kroppenstedtii model is done\n",
      "Corynebacterium matruchotii model is done\n",
      "Corynebacterium otitidis model is done\n",
      "Corynebacterium propinquum model is done\n",
      "Corynebacterium pseudogenitalium model is done\n",
      "Corynebacterium resistens model is done\n",
      "Corynebacterium striatum model is done\n",
      "Corynebacterium tuberculostearicum model is done\n",
      "Corynebacterium urealyticum model is done\n",
      "Cutibacterium acnes model is done\n",
      "Cutibacterium granulosum model is done\n",
      "Deinococcus geothermalis model is done\n",
      "Deinococcus wulumuqiensis model is done\n",
      "Delftia acidovorans model is done\n",
      "Dermacoccus sp. model is done\n",
      "Dietzia cinnamea model is done\n",
      "Elizabethkingia anophelis model is done\n",
      "Enhydrobacter aerosaccus model is done\n",
      "Enterobacter cloacae model is done\n",
      "Enterococcus faecalis model is done\n",
      "Enterococcus faecium model is done\n",
      "Escherichia coli model is done\n",
      "Finegoldia magna model is done\n",
      "Gardnerella vaginalis model is done\n",
      "Gordonia terrae model is done\n",
      "Haemophilus parainfluenzae model is done\n",
      "Hydrocarboniphaga effusa model is done\n",
      "Janibacter hoylei model is done\n",
      "Klebsiella pneumoniae model is done\n",
      "Kocuria atrinae model is done\n",
      "Kocuria rhizophila model is done\n",
      "Kocuria sp. model is done\n",
      "Kytococcus sedentarius model is done\n",
      "Lactococcus lactis model is done\n",
      "Lautropia mirabilis model is done\n",
      "Leuconostoc lactis model is done\n",
      "Limosilactobacillus fermentum model is done\n",
      "Massilia timonae model is done\n",
      "Mediterraneibacter gnavus model is done\n",
      "Methylobacterium radiotolerans model is done\n",
      "Methylorubrum extorquens model is done\n",
      "Methylorubrum populi model is done\n",
      "Microbacterium laevaniformans model is done\n",
      "Micrococcus luteus model is done\n",
      "Mycolicibacterium fortuitum model is done\n",
      "Neisseria flavescens model is done\n",
      "Neisseria sicca model is done\n",
      "Perlucidibaca piscinae model is done\n",
      "Phocaeicola dorei model is done\n",
      "Prevotella bivia model is done\n",
      "Prevotella histicola model is done\n",
      "Prevotella melaninogenica model is done\n",
      "Proteus mirabilis model is done\n",
      "Pseudomonas aeruginosa model is done\n",
      "Pseudomonas putida model is done\n",
      "Ralstonia pickettii model is done\n",
      "Roseomonas sp. model is done\n",
      "Rothia aeria model is done\n",
      "Rothia dentocariosa model is done\n",
      "Rothia mucilaginosa model is done\n",
      "Scardovia wiggsiae model is done\n",
      "Serinicoccus profundi model is done\n",
      "Serratia marcescens model is done\n",
      "Sphingobium yanoikuyae model is done\n",
      "Sphingomonas echinoides model is done\n",
      "Staphylococcus aureus model is done\n",
      "Staphylococcus capitis model is done\n",
      "Staphylococcus epidermidis model is done\n",
      "Staphylococcus haemolyticus model is done\n",
      "Staphylococcus hominis model is done\n",
      "Staphylococcus pettenkoferi model is done\n",
      "Staphylococcus warneri model is done\n",
      "Stenotrophomonas maltophilia model is done\n",
      "Streptococcus anginosus model is done\n",
      "Streptococcus australis model is done\n",
      "Streptococcus gordonii model is done\n",
      "Streptococcus infantis model is done\n",
      "Streptococcus parasanguinis model is done\n",
      "Streptococcus salivarius model is done\n",
      "Streptococcus sanguinis model is done\n",
      "Stutzerimonas stutzeri model is done\n",
      "Veillonella atypica model is done\n",
      "Veillonella parvula model is done\n",
      "Xanthomonas citri model is done\n",
      "[Ruminococcus] torques model is done\n"
     ]
    }
   ],
   "source": [
    "# Run the bash script\n",
    "os.chdir(model_path)\n",
    "\n",
    "!chmod +x {carve_file_path}\n",
    "!{carve_file_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>gram_staining</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organism_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [assembly_accession, gram_staining]\n",
       "Index: []"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check all generated models\n",
    "models = glob.glob(model_path + '*.xml')\n",
    "models = [item.replace(model_path, '') for item in models]\n",
    "models = [item.replace('.xml', '') for item in models]\n",
    "models = [item.replace('_', ' ') for item in models]\n",
    "\n",
    "models\n",
    "\n",
    "# Check non-generated models\n",
    "model_ng = list(set(genome_info.index) - set(models))\n",
    "model_ng\n",
    "\n",
    "# Print accession for the genomes\n",
    "missing_genome_info = genome_info.loc[model_ng]\n",
    "missing_genome_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
