{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------------------------------------------\n",
    "# Script: 5 (Distances.ipynb)\n",
    "# ------------------------------------------\n",
    "# Author: Pratyay Sengupta\n",
    "# ------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from io import StringIO\n",
    "from Bio import Phylo\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from reframed import load_cbmodel\n",
    "from itertools import combinations\n",
    "\n",
    "\n",
    "# Set data directories\n",
    "input_dir = '/mnt/Local_Disk_1/2_Hospital_Microbiome/Data/Modeling/Annotated_genomes/'\n",
    "output_dir = '/mnt/Local_Disk_1/2_Hospital_Microbiome/Data/Output_data/'\n",
    "model_dir = '/mnt/Local_Disk_1/2_Hospital_Microbiome/Data/Modeling/Models/'\n",
    "figures_dir = '/mnt/Local_Disk_1/2_Hospital_Microbiome/Data/Figures/'\n",
    "dist_path = os.path.join(output_dir, 'distances')\n",
    "os.makedirs(dist_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------------------------------\n",
    "# Genome â†’ Organism Mapping\n",
    "# ------------------------------\n",
    "genome_to_organism = pd.read_csv(os.path.join(output_dir, '3_accession_details.csv'), index_col=0)\n",
    "mapping = genome_to_organism['assembly_accession'].to_dict()\n",
    "mapping_inv = {v: k for k, v in mapping.items()}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GCF_037041345.1': 'Abiotrophia defectiva',\n",
       " 'GCF_022870085.1': 'Achromobacter xylosoxidans',\n",
       " 'GCF_000369385.1': 'Acinetobacter baumannii',\n",
       " 'GCF_016027055.1': 'Acinetobacter johnsonii',\n",
       " 'GCF_000430225.1': 'Acinetobacter junii',\n",
       " 'GCF_019048305.1': 'Acinetobacter lwoffii',\n",
       " 'GCF_000248155.1': 'Acinetobacter parvus',\n",
       " 'GCF_000368625.1': 'Acinetobacter schindleri',\n",
       " 'GCF_000248135.1': 'Acinetobacter ursingii',\n",
       " 'GCF_016127955.1': 'Actinomyces oris',\n",
       " 'GCF_900637975.1': 'Actinomyces viscosus',\n",
       " 'GCF_900445095.1': 'Aerococcus viridans',\n",
       " 'GCF_022453665.1': 'Bacteroides thetaiotaomicron',\n",
       " 'GCF_016117815.1': 'Bacteroides uniformis',\n",
       " 'GCF_000741415.1': 'Bifidobacterium adolescentis',\n",
       " 'GCF_000158015.1': 'Bifidobacterium breve',\n",
       " 'GCF_017132775.1': 'Bifidobacterium longum',\n",
       " 'GCF_016907455.1': 'Brachybacterium muris',\n",
       " 'GCF_003994255.1': 'Brachybacterium paraconglomeratum',\n",
       " 'GCF_900169275.1': 'Brevibacterium casei',\n",
       " 'GCF_001584405.1': 'Brevibacterium ravenspurgense',\n",
       " 'GCF_000204035.1': 'Brevundimonas diminuta',\n",
       " 'GCF_002858865.1': 'Caulobacter vibrioides',\n",
       " 'GCF_000317125.1': 'Chroococcidiopsis thermalis',\n",
       " 'GCF_000169035.1': 'Collinsella aerofaciens',\n",
       " 'GCF_023520795.1': 'Corynebacterium accolens',\n",
       " 'GCF_030408675.1': 'Corynebacterium durum',\n",
       " 'GCF_900176155.1': 'Corynebacterium glucuronolyticum',\n",
       " 'GCF_000163435.1': 'Corynebacterium jeikeium',\n",
       " 'GCF_000023145.1': 'Corynebacterium kroppenstedtii',\n",
       " 'GCF_900447555.1': 'Corynebacterium matruchotii',\n",
       " 'GCF_000297795.2': 'Corynebacterium otitidis',\n",
       " 'GCF_016728665.1': 'Corynebacterium propinquum',\n",
       " 'GCF_024453815.1': 'Corynebacterium pseudogenitalium',\n",
       " 'GCF_000177535.2': 'Corynebacterium resistens',\n",
       " 'GCF_016403285.1': 'Corynebacterium striatum',\n",
       " 'GCF_016728365.1': 'Corynebacterium tuberculostearicum',\n",
       " 'GCF_000069945.1': 'Corynebacterium urealyticum',\n",
       " 'GCF_000472405.1': 'Cutibacterium acnes',\n",
       " 'GCF_000463665.1': 'Cutibacterium granulosum',\n",
       " 'GCF_000196275.1': 'Deinococcus geothermalis',\n",
       " 'GCF_014645575.1': 'Deinococcus wulumuqiensis',\n",
       " 'GCF_000741825.1': 'Delftia acidovorans',\n",
       " 'GCF_000152185.1': 'Dermacoccus sp. Ellin185',\n",
       " 'GCF_001571065.1': 'Dietzia cinnamea',\n",
       " 'GCF_000331815.1': 'Elizabethkingia anophelis',\n",
       " 'GCF_900167455.1': 'Enhydrobacter aerosaccus',\n",
       " 'GCF_019047105.1': 'Enterobacter cloacae',\n",
       " 'GCF_000392875.1': 'Enterococcus faecalis',\n",
       " 'GCF_026953455.1': 'Enterococcus faecium',\n",
       " 'GCF_003697165.2': 'Escherichia coli',\n",
       " 'GCF_003182075.1': 'Finegoldia magna',\n",
       " 'GCF_000178355.1': 'Gardnerella vaginalis',\n",
       " 'GCF_901542405.1': 'Gordonia terrae',\n",
       " 'GCF_000746485.1': 'Haemophilus parainfluenzae',\n",
       " 'GCF_000271305.1': 'Hydrocarboniphaga effusa',\n",
       " 'GCF_000297495.1': 'Janibacter hoylei',\n",
       " 'GCF_000742135.1': 'Klebsiella pneumoniae',\n",
       " 'GCF_000286355.1': 'Kocuria atrinae',\n",
       " 'GCF_003667225.1': 'Kocuria rhizophila',\n",
       " 'GCF_000349605.1': 'Kocuria sp. UCD-OTCP',\n",
       " 'GCF_000023925.1': 'Kytococcus sedentarius',\n",
       " 'GCF_900099625.1': 'Lactococcus lactis',\n",
       " 'GCF_000186425.1': 'Lautropia mirabilis',\n",
       " 'GCF_000179875.1': 'Leuconostoc lactis',\n",
       " 'GCF_029823225.1': 'Limosilactobacillus fermentum',\n",
       " 'GCF_000315425.1': 'Massilia timonae',\n",
       " 'GCF_008121495.1': 'Mediterraneibacter gnavus',\n",
       " 'GCF_000019725.1': 'Methylobacterium radiotolerans',\n",
       " 'GCF_000083545.1': 'Methylorubrum extorquens',\n",
       " 'GCF_000019945.1': 'Methylorubrum populi',\n",
       " 'GCF_016907555.1': 'Microbacterium laevaniformans',\n",
       " 'GCF_006094415.1': 'Micrococcus luteus',\n",
       " 'GCF_022179545.1': 'Mycolicibacterium fortuitum',\n",
       " 'GCF_900453805.1': 'Neisseria flavescens',\n",
       " 'GCF_000174655.1': 'Neisseria sicca',\n",
       " 'GCF_000420045.1': 'Perlucidibaca piscinae',\n",
       " 'GCF_000156075.1': 'Phocaeicola dorei',\n",
       " 'GCF_000262545.1': 'Prevotella bivia',\n",
       " 'GCF_018128125.1': 'Prevotella histicola',\n",
       " 'GCF_000144405.1': 'Prevotella melaninogenica',\n",
       " 'GCF_000160755.1': 'Proteus mirabilis',\n",
       " 'GCF_024169805.1': 'Pseudomonas aeruginosa',\n",
       " 'GCF_013387005.1': 'Pseudomonas putida',\n",
       " 'GCF_016466415.2': 'Ralstonia pickettii',\n",
       " 'GCF_000292225.1': 'Roseomonas sp. B5',\n",
       " 'GCF_016726365.1': 'Rothia aeria',\n",
       " 'GCF_000164695.2': 'Rothia dentocariosa',\n",
       " 'GCF_000175615.1': 'Rothia mucilaginosa',\n",
       " 'GCF_000275805.1': 'Scardovia wiggsiae',\n",
       " 'GCF_000224715.1': 'Serinicoccus profundi',\n",
       " 'GCF_034424255.1': 'Serratia marcescens',\n",
       " 'GCF_034424525.1': 'Sphingobium yanoikuyae',\n",
       " 'GCF_033842725.1': 'Sphingomonas echinoides',\n",
       " 'GCF_001027105.1': 'Staphylococcus aureus',\n",
       " 'GCF_014635765.1': 'Staphylococcus capitis',\n",
       " 'GCF_006094375.1': 'Staphylococcus epidermidis',\n",
       " 'GCF_029023945.1': 'Staphylococcus haemolyticus',\n",
       " 'GCF_002902465.1': 'Staphylococcus hominis',\n",
       " 'GCF_016127935.1': 'Staphylococcus pettenkoferi',\n",
       " 'GCF_029023845.1': 'Staphylococcus warneri',\n",
       " 'GCF_000597745.1': 'Stenotrophomonas maltophilia',\n",
       " 'GCF_000257765.1': 'Streptococcus anginosus',\n",
       " 'GCF_000222745.1': 'Streptococcus australis',\n",
       " 'GCF_019046945.1': 'Streptococcus gordonii',\n",
       " 'GCF_000187465.1': 'Streptococcus infantis',\n",
       " 'GCF_900459355.1': 'Streptococcus parasanguinis',\n",
       " 'GCF_031192325.1': 'Streptococcus salivarius',\n",
       " 'GCF_001593525.1': 'Streptococcus sanguinis',\n",
       " 'GCF_000219605.1': 'Stutzerimonas stutzeri',\n",
       " 'GCF_002959915.1': 'Veillonella atypica',\n",
       " 'GCF_000024945.1': 'Veillonella parvula',\n",
       " 'GCF_002018575.1': 'Xanthomonas citri',\n",
       " 'GCF_000153925.1': '[Ruminococcus] torques'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_inv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Classify the genomes using COG classifier (Alternatively command-line can be used)\n",
    "#os.chdir(input_path)\n",
    "\n",
    "#for genomes in glob.glob('*.faa'):\n",
    "#    outfile = output_path + 'COG_classifier/' + genomes[:-4] \n",
    "#    cogclassifier.run(genomes, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GCF_000019725.1</th>\n",
       "      <th>GCF_000019945.1</th>\n",
       "      <th>GCF_000023145.1</th>\n",
       "      <th>GCF_000023925.1</th>\n",
       "      <th>GCF_000024945.1</th>\n",
       "      <th>GCF_000069945.1</th>\n",
       "      <th>GCF_000083545.1</th>\n",
       "      <th>GCF_000144405.1</th>\n",
       "      <th>GCF_000152185.1</th>\n",
       "      <th>GCF_000153925.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000019725.1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2078</td>\n",
       "      <td>0.6516</td>\n",
       "      <td>0.6056</td>\n",
       "      <td>0.6612</td>\n",
       "      <td>0.6507</td>\n",
       "      <td>0.2194</td>\n",
       "      <td>0.7032</td>\n",
       "      <td>0.5773</td>\n",
       "      <td>0.6998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000019945.1</th>\n",
       "      <td>0.2078</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6450</td>\n",
       "      <td>0.6067</td>\n",
       "      <td>0.6615</td>\n",
       "      <td>0.6473</td>\n",
       "      <td>0.1162</td>\n",
       "      <td>0.6929</td>\n",
       "      <td>0.5821</td>\n",
       "      <td>0.6894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000023145.1</th>\n",
       "      <td>0.6516</td>\n",
       "      <td>0.6450</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4675</td>\n",
       "      <td>0.5666</td>\n",
       "      <td>0.3217</td>\n",
       "      <td>0.6406</td>\n",
       "      <td>0.6284</td>\n",
       "      <td>0.4338</td>\n",
       "      <td>0.6265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000023925.1</th>\n",
       "      <td>0.6056</td>\n",
       "      <td>0.6067</td>\n",
       "      <td>0.4675</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6163</td>\n",
       "      <td>0.4411</td>\n",
       "      <td>0.5949</td>\n",
       "      <td>0.6506</td>\n",
       "      <td>0.3516</td>\n",
       "      <td>0.6404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000024945.1</th>\n",
       "      <td>0.6612</td>\n",
       "      <td>0.6615</td>\n",
       "      <td>0.5666</td>\n",
       "      <td>0.6163</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5866</td>\n",
       "      <td>0.6599</td>\n",
       "      <td>0.5947</td>\n",
       "      <td>0.5929</td>\n",
       "      <td>0.5329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000069945.1</th>\n",
       "      <td>0.6507</td>\n",
       "      <td>0.6473</td>\n",
       "      <td>0.3217</td>\n",
       "      <td>0.4411</td>\n",
       "      <td>0.5866</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6409</td>\n",
       "      <td>0.6163</td>\n",
       "      <td>0.4297</td>\n",
       "      <td>0.6164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000083545.1</th>\n",
       "      <td>0.2194</td>\n",
       "      <td>0.1162</td>\n",
       "      <td>0.6406</td>\n",
       "      <td>0.5949</td>\n",
       "      <td>0.6599</td>\n",
       "      <td>0.6409</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6888</td>\n",
       "      <td>0.5748</td>\n",
       "      <td>0.6898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000144405.1</th>\n",
       "      <td>0.7032</td>\n",
       "      <td>0.6929</td>\n",
       "      <td>0.6284</td>\n",
       "      <td>0.6506</td>\n",
       "      <td>0.5947</td>\n",
       "      <td>0.6163</td>\n",
       "      <td>0.6888</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6414</td>\n",
       "      <td>0.6144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000152185.1</th>\n",
       "      <td>0.5773</td>\n",
       "      <td>0.5821</td>\n",
       "      <td>0.4338</td>\n",
       "      <td>0.3516</td>\n",
       "      <td>0.5929</td>\n",
       "      <td>0.4297</td>\n",
       "      <td>0.5748</td>\n",
       "      <td>0.6414</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000153925.1</th>\n",
       "      <td>0.6998</td>\n",
       "      <td>0.6894</td>\n",
       "      <td>0.6265</td>\n",
       "      <td>0.6404</td>\n",
       "      <td>0.5329</td>\n",
       "      <td>0.6164</td>\n",
       "      <td>0.6898</td>\n",
       "      <td>0.6144</td>\n",
       "      <td>0.6272</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GCF_000019725.1  GCF_000019945.1  GCF_000023145.1  \\\n",
       "GCF_000019725.1           0.0000           0.2078           0.6516   \n",
       "GCF_000019945.1           0.2078           0.0000           0.6450   \n",
       "GCF_000023145.1           0.6516           0.6450           0.0000   \n",
       "GCF_000023925.1           0.6056           0.6067           0.4675   \n",
       "GCF_000024945.1           0.6612           0.6615           0.5666   \n",
       "GCF_000069945.1           0.6507           0.6473           0.3217   \n",
       "GCF_000083545.1           0.2194           0.1162           0.6406   \n",
       "GCF_000144405.1           0.7032           0.6929           0.6284   \n",
       "GCF_000152185.1           0.5773           0.5821           0.4338   \n",
       "GCF_000153925.1           0.6998           0.6894           0.6265   \n",
       "\n",
       "                 GCF_000023925.1  GCF_000024945.1  GCF_000069945.1  \\\n",
       "GCF_000019725.1           0.6056           0.6612           0.6507   \n",
       "GCF_000019945.1           0.6067           0.6615           0.6473   \n",
       "GCF_000023145.1           0.4675           0.5666           0.3217   \n",
       "GCF_000023925.1           0.0000           0.6163           0.4411   \n",
       "GCF_000024945.1           0.6163           0.0000           0.5866   \n",
       "GCF_000069945.1           0.4411           0.5866           0.0000   \n",
       "GCF_000083545.1           0.5949           0.6599           0.6409   \n",
       "GCF_000144405.1           0.6506           0.5947           0.6163   \n",
       "GCF_000152185.1           0.3516           0.5929           0.4297   \n",
       "GCF_000153925.1           0.6404           0.5329           0.6164   \n",
       "\n",
       "                 GCF_000083545.1  GCF_000144405.1  GCF_000152185.1  \\\n",
       "GCF_000019725.1           0.2194           0.7032           0.5773   \n",
       "GCF_000019945.1           0.1162           0.6929           0.5821   \n",
       "GCF_000023145.1           0.6406           0.6284           0.4338   \n",
       "GCF_000023925.1           0.5949           0.6506           0.3516   \n",
       "GCF_000024945.1           0.6599           0.5947           0.5929   \n",
       "GCF_000069945.1           0.6409           0.6163           0.4297   \n",
       "GCF_000083545.1           0.0000           0.6888           0.5748   \n",
       "GCF_000144405.1           0.6888           0.0000           0.6414   \n",
       "GCF_000152185.1           0.5748           0.6414           0.0000   \n",
       "GCF_000153925.1           0.6898           0.6144           0.6272   \n",
       "\n",
       "                 GCF_000153925.1  \n",
       "GCF_000019725.1           0.6998  \n",
       "GCF_000019945.1           0.6894  \n",
       "GCF_000023145.1           0.6265  \n",
       "GCF_000023925.1           0.6404  \n",
       "GCF_000024945.1           0.5329  \n",
       "GCF_000069945.1           0.6164  \n",
       "GCF_000083545.1           0.6898  \n",
       "GCF_000144405.1           0.6144  \n",
       "GCF_000152185.1           0.6272  \n",
       "GCF_000153925.1           0.0000  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------\n",
    "# Functional Distance from COG Annotations\n",
    "# ------------------------------\n",
    "cog_dir = os.path.join(output_dir, 'COG_classifier')\n",
    "cog_dict = {}\n",
    "genomes = [g for g in os.listdir(cog_dir) if os.path.isdir(os.path.join(cog_dir, g))]\n",
    "\n",
    "for genome in genomes:\n",
    "    cog_file = os.path.join(cog_dir, genome, 'classifier_result.tsv')\n",
    "    if os.path.exists(cog_file):\n",
    "        cog_data = pd.read_csv(cog_file, sep='\\t')\n",
    "        filtered = cog_data[cog_data['IDENTITY'] > 30]['COG_ID']\n",
    "        for cog in filtered:\n",
    "            cog_dict.setdefault(cog, {})[genome] = 1\n",
    "\n",
    "# Fill absent values with 0\n",
    "for cog in cog_dict:\n",
    "    for genome in genomes:\n",
    "        cog_dict[cog].setdefault(genome, 0)\n",
    "\n",
    "cog_df = pd.DataFrame(cog_dict).T.fillna(0).sort_index(axis=0).sort_index(axis=1)\n",
    "\n",
    "# Jaccard distance\n",
    "cog_dist = pdist(cog_df.T, metric='jaccard')\n",
    "cog_matrix = squareform(cog_dist)\n",
    "cog_distance_df = pd.DataFrame(cog_matrix, index=cog_df.columns, columns=cog_df.columns).round(4)\n",
    "\n",
    "cog_distance_df.to_csv(os.path.join(dist_path, '1_functional.tsv'), sep='\\t')\n",
    "\n",
    "# Display the distance matrix\n",
    "cog_distance_df.iloc[0:10, 0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculating the mash distance in the command line\n",
    "# mash sketch -o genome_sketch ./../../Genomes/*.fa\n",
    "# mash dist genome_sketch.msh genome_sketch.msh > mash_distance.tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Query</th>\n",
       "      <th>GCF_000019725.1</th>\n",
       "      <th>GCF_000019945.1</th>\n",
       "      <th>GCF_000023145.1</th>\n",
       "      <th>GCF_000023925.1</th>\n",
       "      <th>GCF_000024945.1</th>\n",
       "      <th>GCF_000069945.1</th>\n",
       "      <th>GCF_000083545.1</th>\n",
       "      <th>GCF_000144405.1</th>\n",
       "      <th>GCF_000152185.1</th>\n",
       "      <th>GCF_000153925.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000019725.1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1782</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.1823</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000019945.1</th>\n",
       "      <td>0.1782</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.1045</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.296</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000023145.1</th>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.296</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000023925.1</th>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.296</td>\n",
       "      <td>0.2960</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000024945.1</th>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000069945.1</th>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.296</td>\n",
       "      <td>0.296</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000083545.1</th>\n",
       "      <td>0.1823</td>\n",
       "      <td>0.1045</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.296</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000144405.1</th>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000152185.1</th>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.2960</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000153925.1</th>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.296</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Query            GCF_000019725.1  GCF_000019945.1  GCF_000023145.1  \\\n",
       "Reference                                                            \n",
       "GCF_000019725.1           0.0000           0.1782            1.000   \n",
       "GCF_000019945.1           0.1782           0.0000            1.000   \n",
       "GCF_000023145.1           1.0000           1.0000            0.000   \n",
       "GCF_000023925.1           1.0000           1.0000            1.000   \n",
       "GCF_000024945.1           1.0000           1.0000            1.000   \n",
       "GCF_000069945.1           1.0000           1.0000            0.296   \n",
       "GCF_000083545.1           0.1823           0.1045            1.000   \n",
       "GCF_000144405.1           1.0000           1.0000            1.000   \n",
       "GCF_000152185.1           1.0000           0.2960            1.000   \n",
       "GCF_000153925.1           1.0000           1.0000            1.000   \n",
       "\n",
       "Query            GCF_000023925.1  GCF_000024945.1  GCF_000069945.1  \\\n",
       "Reference                                                            \n",
       "GCF_000019725.1            1.000            1.000            1.000   \n",
       "GCF_000019945.1            1.000            1.000            1.000   \n",
       "GCF_000023145.1            1.000            1.000            0.296   \n",
       "GCF_000023925.1            0.000            1.000            0.296   \n",
       "GCF_000024945.1            1.000            0.000            1.000   \n",
       "GCF_000069945.1            0.296            1.000            0.000   \n",
       "GCF_000083545.1            0.296            1.000            1.000   \n",
       "GCF_000144405.1            1.000            1.000            1.000   \n",
       "GCF_000152185.1            1.000            1.000            1.000   \n",
       "GCF_000153925.1            1.000            0.296            1.000   \n",
       "\n",
       "Query            GCF_000083545.1  GCF_000144405.1  GCF_000152185.1  \\\n",
       "Reference                                                            \n",
       "GCF_000019725.1           0.1823              1.0            1.000   \n",
       "GCF_000019945.1           0.1045              1.0            0.296   \n",
       "GCF_000023145.1           1.0000              1.0            1.000   \n",
       "GCF_000023925.1           0.2960              1.0            1.000   \n",
       "GCF_000024945.1           1.0000              1.0            1.000   \n",
       "GCF_000069945.1           1.0000              1.0            1.000   \n",
       "GCF_000083545.1           0.0000              1.0            1.000   \n",
       "GCF_000144405.1           1.0000              0.0            1.000   \n",
       "GCF_000152185.1           1.0000              1.0            0.000   \n",
       "GCF_000153925.1           1.0000              1.0            1.000   \n",
       "\n",
       "Query            GCF_000153925.1  \n",
       "Reference                         \n",
       "GCF_000019725.1            1.000  \n",
       "GCF_000019945.1            1.000  \n",
       "GCF_000023145.1            1.000  \n",
       "GCF_000023925.1            1.000  \n",
       "GCF_000024945.1            0.296  \n",
       "GCF_000069945.1            1.000  \n",
       "GCF_000083545.1            1.000  \n",
       "GCF_000144405.1            1.000  \n",
       "GCF_000152185.1            1.000  \n",
       "GCF_000153925.1            0.000  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------\n",
    "# MASH Distance\n",
    "# ------------------------------\n",
    "mash_file = os.path.join(output_dir, 'mash_distance', 'mash_distance.tab')\n",
    "mash_df = pd.read_csv(mash_file, sep='\\t', header=None,\n",
    "                      names=['Reference', 'Query', 'mash_distance', 'p-value', 'shared_hashes'])\n",
    "\n",
    "# Clean labels\n",
    "clean_label = lambda x: os.path.basename(x).replace('.fa', '')\n",
    "mash_df['Reference'] = mash_df['Reference'].apply(clean_label)\n",
    "mash_df['Query'] = mash_df['Query'].apply(clean_label)\n",
    "\n",
    "mash_matrix = mash_df.pivot(index='Reference', columns='Query', values='mash_distance').sort_index().sort_index(axis=1).round(4)\n",
    "mash_matrix.to_csv(os.path.join(dist_path, '2_mash.tsv'), sep='\\t')\n",
    "\n",
    "# Display the distance matrix\n",
    "mash_matrix.iloc[0:10, 0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculating the phylogenetic distance and tree using GToTree in the command line\n",
    "# GToTree -f ./../../Genomes/fasta_files.txt -H Bacteria -t -L Species,Strain -j 10 -o Hosp_microbiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GCF_000019725.1</th>\n",
       "      <th>GCF_000019945.1</th>\n",
       "      <th>GCF_000023145.1</th>\n",
       "      <th>GCF_000023925.1</th>\n",
       "      <th>GCF_000024945.1</th>\n",
       "      <th>GCF_000069945.1</th>\n",
       "      <th>GCF_000083545.1</th>\n",
       "      <th>GCF_000144405.1</th>\n",
       "      <th>GCF_000152185.1</th>\n",
       "      <th>GCF_000153925.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_000019725.1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1192</td>\n",
       "      <td>1.5036</td>\n",
       "      <td>1.5570</td>\n",
       "      <td>1.3167</td>\n",
       "      <td>1.5327</td>\n",
       "      <td>0.1208</td>\n",
       "      <td>1.6713</td>\n",
       "      <td>1.5205</td>\n",
       "      <td>1.4402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000019945.1</th>\n",
       "      <td>0.1192</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.5106</td>\n",
       "      <td>1.5640</td>\n",
       "      <td>1.3237</td>\n",
       "      <td>1.5398</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>1.6784</td>\n",
       "      <td>1.5275</td>\n",
       "      <td>1.4473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000023145.1</th>\n",
       "      <td>1.5036</td>\n",
       "      <td>1.5106</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.8307</td>\n",
       "      <td>1.2721</td>\n",
       "      <td>0.3165</td>\n",
       "      <td>1.5122</td>\n",
       "      <td>1.7149</td>\n",
       "      <td>0.7942</td>\n",
       "      <td>1.3957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000023925.1</th>\n",
       "      <td>1.5570</td>\n",
       "      <td>1.5640</td>\n",
       "      <td>0.8307</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.3256</td>\n",
       "      <td>0.8599</td>\n",
       "      <td>1.5656</td>\n",
       "      <td>1.7683</td>\n",
       "      <td>0.4279</td>\n",
       "      <td>1.4491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000024945.1</th>\n",
       "      <td>1.3167</td>\n",
       "      <td>1.3237</td>\n",
       "      <td>1.2721</td>\n",
       "      <td>1.3256</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.3013</td>\n",
       "      <td>1.3253</td>\n",
       "      <td>1.5280</td>\n",
       "      <td>1.2891</td>\n",
       "      <td>1.0730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000069945.1</th>\n",
       "      <td>1.5327</td>\n",
       "      <td>1.5398</td>\n",
       "      <td>0.3165</td>\n",
       "      <td>0.8599</td>\n",
       "      <td>1.3013</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.5414</td>\n",
       "      <td>1.7441</td>\n",
       "      <td>0.8234</td>\n",
       "      <td>1.4248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000083545.1</th>\n",
       "      <td>0.1208</td>\n",
       "      <td>0.0300</td>\n",
       "      <td>1.5122</td>\n",
       "      <td>1.5656</td>\n",
       "      <td>1.3253</td>\n",
       "      <td>1.5414</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.6800</td>\n",
       "      <td>1.5292</td>\n",
       "      <td>1.4489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000144405.1</th>\n",
       "      <td>1.6713</td>\n",
       "      <td>1.6784</td>\n",
       "      <td>1.7149</td>\n",
       "      <td>1.7683</td>\n",
       "      <td>1.5280</td>\n",
       "      <td>1.7441</td>\n",
       "      <td>1.6800</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.7318</td>\n",
       "      <td>1.6516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000152185.1</th>\n",
       "      <td>1.5205</td>\n",
       "      <td>1.5275</td>\n",
       "      <td>0.7942</td>\n",
       "      <td>0.4279</td>\n",
       "      <td>1.2891</td>\n",
       "      <td>0.8234</td>\n",
       "      <td>1.5292</td>\n",
       "      <td>1.7318</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.4126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000153925.1</th>\n",
       "      <td>1.4402</td>\n",
       "      <td>1.4473</td>\n",
       "      <td>1.3957</td>\n",
       "      <td>1.4491</td>\n",
       "      <td>1.0730</td>\n",
       "      <td>1.4248</td>\n",
       "      <td>1.4489</td>\n",
       "      <td>1.6516</td>\n",
       "      <td>1.4126</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GCF_000019725.1  GCF_000019945.1  GCF_000023145.1  \\\n",
       "GCF_000019725.1           0.0000           0.1192           1.5036   \n",
       "GCF_000019945.1           0.1192           0.0000           1.5106   \n",
       "GCF_000023145.1           1.5036           1.5106           0.0000   \n",
       "GCF_000023925.1           1.5570           1.5640           0.8307   \n",
       "GCF_000024945.1           1.3167           1.3237           1.2721   \n",
       "GCF_000069945.1           1.5327           1.5398           0.3165   \n",
       "GCF_000083545.1           0.1208           0.0300           1.5122   \n",
       "GCF_000144405.1           1.6713           1.6784           1.7149   \n",
       "GCF_000152185.1           1.5205           1.5275           0.7942   \n",
       "GCF_000153925.1           1.4402           1.4473           1.3957   \n",
       "\n",
       "                 GCF_000023925.1  GCF_000024945.1  GCF_000069945.1  \\\n",
       "GCF_000019725.1           1.5570           1.3167           1.5327   \n",
       "GCF_000019945.1           1.5640           1.3237           1.5398   \n",
       "GCF_000023145.1           0.8307           1.2721           0.3165   \n",
       "GCF_000023925.1           0.0000           1.3256           0.8599   \n",
       "GCF_000024945.1           1.3256           0.0000           1.3013   \n",
       "GCF_000069945.1           0.8599           1.3013           0.0000   \n",
       "GCF_000083545.1           1.5656           1.3253           1.5414   \n",
       "GCF_000144405.1           1.7683           1.5280           1.7441   \n",
       "GCF_000152185.1           0.4279           1.2891           0.8234   \n",
       "GCF_000153925.1           1.4491           1.0730           1.4248   \n",
       "\n",
       "                 GCF_000083545.1  GCF_000144405.1  GCF_000152185.1  \\\n",
       "GCF_000019725.1           0.1208           1.6713           1.5205   \n",
       "GCF_000019945.1           0.0300           1.6784           1.5275   \n",
       "GCF_000023145.1           1.5122           1.7149           0.7942   \n",
       "GCF_000023925.1           1.5656           1.7683           0.4279   \n",
       "GCF_000024945.1           1.3253           1.5280           1.2891   \n",
       "GCF_000069945.1           1.5414           1.7441           0.8234   \n",
       "GCF_000083545.1           0.0000           1.6800           1.5292   \n",
       "GCF_000144405.1           1.6800           0.0000           1.7318   \n",
       "GCF_000152185.1           1.5292           1.7318           0.0000   \n",
       "GCF_000153925.1           1.4489           1.6516           1.4126   \n",
       "\n",
       "                 GCF_000153925.1  \n",
       "GCF_000019725.1           1.4402  \n",
       "GCF_000019945.1           1.4473  \n",
       "GCF_000023145.1           1.3957  \n",
       "GCF_000023925.1           1.4491  \n",
       "GCF_000024945.1           1.0730  \n",
       "GCF_000069945.1           1.4248  \n",
       "GCF_000083545.1           1.4489  \n",
       "GCF_000144405.1           1.6516  \n",
       "GCF_000152185.1           1.4126  \n",
       "GCF_000153925.1           0.0000  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------\n",
    "# Phylogenetic Distance from Tree\n",
    "# ------------------------------\n",
    "newick_file = os.path.join(output_dir, 'phylogenetic_tree', 'Hosp_microbiome', 'Hosp_microbiome.tre')\n",
    "tree = Phylo.read(newick_file, 'newick')\n",
    "\n",
    "def compute_phylo_matrix(tree):\n",
    "    terms = tree.get_terminals()\n",
    "    labels = [term.name for term in terms]\n",
    "    matrix = [[tree.distance(i, j) for j in terms] for i in terms]\n",
    "    return pd.DataFrame(matrix, index=labels, columns=labels)\n",
    "\n",
    "phylo_df = compute_phylo_matrix(tree).sort_index().sort_index(axis=1).round(4)\n",
    "phylo_df.to_csv(os.path.join(dist_path, '3_phylogenetic.tsv'), sep='\\t')\n",
    "\n",
    "# Display the distance matrix\n",
    "phylo_df.iloc[0:10, 0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GCF_037041345.1</th>\n",
       "      <th>GCF_022870085.1</th>\n",
       "      <th>GCF_000369385.1</th>\n",
       "      <th>GCF_016027055.1</th>\n",
       "      <th>GCF_000430225.1</th>\n",
       "      <th>GCF_019048305.1</th>\n",
       "      <th>GCF_000248155.1</th>\n",
       "      <th>GCF_000368625.1</th>\n",
       "      <th>GCF_000248135.1</th>\n",
       "      <th>GCF_016127955.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCF_037041345.1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.7977</td>\n",
       "      <td>0.7599</td>\n",
       "      <td>0.7721</td>\n",
       "      <td>0.7571</td>\n",
       "      <td>0.7639</td>\n",
       "      <td>0.7682</td>\n",
       "      <td>0.7740</td>\n",
       "      <td>0.7679</td>\n",
       "      <td>0.6162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_022870085.1</th>\n",
       "      <td>0.7977</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5430</td>\n",
       "      <td>0.5843</td>\n",
       "      <td>0.5915</td>\n",
       "      <td>0.5706</td>\n",
       "      <td>0.6371</td>\n",
       "      <td>0.5719</td>\n",
       "      <td>0.5713</td>\n",
       "      <td>0.6805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000369385.1</th>\n",
       "      <td>0.7599</td>\n",
       "      <td>0.5430</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3904</td>\n",
       "      <td>0.3836</td>\n",
       "      <td>0.3920</td>\n",
       "      <td>0.4471</td>\n",
       "      <td>0.4182</td>\n",
       "      <td>0.3800</td>\n",
       "      <td>0.6695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_016027055.1</th>\n",
       "      <td>0.7721</td>\n",
       "      <td>0.5843</td>\n",
       "      <td>0.3904</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2631</td>\n",
       "      <td>0.2653</td>\n",
       "      <td>0.3081</td>\n",
       "      <td>0.2964</td>\n",
       "      <td>0.2798</td>\n",
       "      <td>0.6611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000430225.1</th>\n",
       "      <td>0.7571</td>\n",
       "      <td>0.5915</td>\n",
       "      <td>0.3836</td>\n",
       "      <td>0.2631</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2934</td>\n",
       "      <td>0.3008</td>\n",
       "      <td>0.3275</td>\n",
       "      <td>0.2809</td>\n",
       "      <td>0.6480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_019048305.1</th>\n",
       "      <td>0.7639</td>\n",
       "      <td>0.5706</td>\n",
       "      <td>0.3920</td>\n",
       "      <td>0.2653</td>\n",
       "      <td>0.2934</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3254</td>\n",
       "      <td>0.2299</td>\n",
       "      <td>0.2946</td>\n",
       "      <td>0.6533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000248155.1</th>\n",
       "      <td>0.7682</td>\n",
       "      <td>0.6371</td>\n",
       "      <td>0.4471</td>\n",
       "      <td>0.3081</td>\n",
       "      <td>0.3008</td>\n",
       "      <td>0.3254</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3524</td>\n",
       "      <td>0.3282</td>\n",
       "      <td>0.6512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000368625.1</th>\n",
       "      <td>0.7740</td>\n",
       "      <td>0.5719</td>\n",
       "      <td>0.4182</td>\n",
       "      <td>0.2964</td>\n",
       "      <td>0.3275</td>\n",
       "      <td>0.2299</td>\n",
       "      <td>0.3524</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2741</td>\n",
       "      <td>0.6795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_000248135.1</th>\n",
       "      <td>0.7679</td>\n",
       "      <td>0.5713</td>\n",
       "      <td>0.3800</td>\n",
       "      <td>0.2798</td>\n",
       "      <td>0.2809</td>\n",
       "      <td>0.2946</td>\n",
       "      <td>0.3282</td>\n",
       "      <td>0.2741</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCF_016127955.1</th>\n",
       "      <td>0.6162</td>\n",
       "      <td>0.6805</td>\n",
       "      <td>0.6695</td>\n",
       "      <td>0.6611</td>\n",
       "      <td>0.6480</td>\n",
       "      <td>0.6533</td>\n",
       "      <td>0.6512</td>\n",
       "      <td>0.6795</td>\n",
       "      <td>0.6595</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GCF_037041345.1  GCF_022870085.1  GCF_000369385.1  \\\n",
       "GCF_037041345.1           0.0000           0.7977           0.7599   \n",
       "GCF_022870085.1           0.7977           0.0000           0.5430   \n",
       "GCF_000369385.1           0.7599           0.5430           0.0000   \n",
       "GCF_016027055.1           0.7721           0.5843           0.3904   \n",
       "GCF_000430225.1           0.7571           0.5915           0.3836   \n",
       "GCF_019048305.1           0.7639           0.5706           0.3920   \n",
       "GCF_000248155.1           0.7682           0.6371           0.4471   \n",
       "GCF_000368625.1           0.7740           0.5719           0.4182   \n",
       "GCF_000248135.1           0.7679           0.5713           0.3800   \n",
       "GCF_016127955.1           0.6162           0.6805           0.6695   \n",
       "\n",
       "                 GCF_016027055.1  GCF_000430225.1  GCF_019048305.1  \\\n",
       "GCF_037041345.1           0.7721           0.7571           0.7639   \n",
       "GCF_022870085.1           0.5843           0.5915           0.5706   \n",
       "GCF_000369385.1           0.3904           0.3836           0.3920   \n",
       "GCF_016027055.1           0.0000           0.2631           0.2653   \n",
       "GCF_000430225.1           0.2631           0.0000           0.2934   \n",
       "GCF_019048305.1           0.2653           0.2934           0.0000   \n",
       "GCF_000248155.1           0.3081           0.3008           0.3254   \n",
       "GCF_000368625.1           0.2964           0.3275           0.2299   \n",
       "GCF_000248135.1           0.2798           0.2809           0.2946   \n",
       "GCF_016127955.1           0.6611           0.6480           0.6533   \n",
       "\n",
       "                 GCF_000248155.1  GCF_000368625.1  GCF_000248135.1  \\\n",
       "GCF_037041345.1           0.7682           0.7740           0.7679   \n",
       "GCF_022870085.1           0.6371           0.5719           0.5713   \n",
       "GCF_000369385.1           0.4471           0.4182           0.3800   \n",
       "GCF_016027055.1           0.3081           0.2964           0.2798   \n",
       "GCF_000430225.1           0.3008           0.3275           0.2809   \n",
       "GCF_019048305.1           0.3254           0.2299           0.2946   \n",
       "GCF_000248155.1           0.0000           0.3524           0.3282   \n",
       "GCF_000368625.1           0.3524           0.0000           0.2741   \n",
       "GCF_000248135.1           0.3282           0.2741           0.0000   \n",
       "GCF_016127955.1           0.6512           0.6795           0.6595   \n",
       "\n",
       "                 GCF_016127955.1  \n",
       "GCF_037041345.1           0.6162  \n",
       "GCF_022870085.1           0.6805  \n",
       "GCF_000369385.1           0.6695  \n",
       "GCF_016027055.1           0.6611  \n",
       "GCF_000430225.1           0.6480  \n",
       "GCF_019048305.1           0.6533  \n",
       "GCF_000248155.1           0.6512  \n",
       "GCF_000368625.1           0.6795  \n",
       "GCF_000248135.1           0.6595  \n",
       "GCF_016127955.1           0.0000  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------\n",
    "# Metabolic Distance (Presence/Absence of Reactions)\n",
    "# ------------------------------\n",
    "reaction_dict = {}\n",
    "models = glob.glob(os.path.join(model_dir, '*.xml'))\n",
    "\n",
    "for path in models:\n",
    "    model_id = os.path.basename(path).replace('.xml', '')\n",
    "    model = load_cbmodel(path)\n",
    "    for rxn in model.reactions:\n",
    "        reaction_dict.setdefault(model_id, {})[rxn] = 1\n",
    "\n",
    "# Fill in 0s for absent reactions\n",
    "all_reactions = set(r for model in reaction_dict.values() for r in model)\n",
    "for model_id in reaction_dict:\n",
    "    for rxn in all_reactions:\n",
    "        reaction_dict[model_id].setdefault(rxn, 0)\n",
    "\n",
    "reaction_df = pd.DataFrame(reaction_dict).T.fillna(0).sort_index().sort_index(axis=1)\n",
    "\n",
    "# Jaccard distance\n",
    "reaction_dist = pdist(reaction_df, metric='jaccard')\n",
    "reaction_matrix = squareform(reaction_dist)\n",
    "reaction_distance_df = pd.DataFrame(reaction_matrix, index=reaction_df.index, columns=reaction_df.index).round(4)\n",
    "\n",
    "# Invert the dictionary: name â†’ accession\n",
    "name_to_accession = {v.replace(' ', '_'): k for k, v in mapping_inv.items()}\n",
    "\n",
    "# Apply specific replacements to keys\n",
    "name_to_accession = {\n",
    "    key.replace('Kocuria_sp._UCD-OTCP', 'Kocuria_sp.')\n",
    "        .replace('Roseomonas_sp._B5', 'Roseomonas_sp.')\n",
    "        .replace('Dermacoccus_sp._Ellin185', 'Dermacoccus_sp.'): val\n",
    "    for key, val in name_to_accession.items()\n",
    "}\n",
    "\n",
    "reaction_distance_df = reaction_distance_df.rename(columns=name_to_accession, index=name_to_accession)\n",
    "\n",
    "reaction_distance_df.to_csv(os.path.join(dist_path, '4_metabolic.tsv'), sep='\\t')\n",
    "\n",
    "# Display the distance matrix\n",
    "reaction_distance_df.iloc[0:10, 0:10]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
